### ISSUE TITLE
Add one-time-code to list of autocomplete attributes

### ISSUE THREAD

**Issue by fstrr** [w3c/wcag#2613]
[WCAG 2.1's list of input control purposes](https://www.w3.org/TR/WCAG21/#input-purposes) lists the relevant `autocomplete` attributes that relate to the user of the content and pertain only to information related to that individual. Included in that list is `current-password` and `new-password`, but [`one-time-code`](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-autocomplete-one-time-code) isn't, but feels like it should be.

**Comment by patrickhlauke** [comment]
> The following input control purposes are intended to relate to the user of the content and pertain only to information related to that individual.

a one-time code isn't really (personal) information related to the user

**Comment by yatil** [comment]
I agree with Patrick here. Maybe one could extend accessible authentication to include a requirement for using `one-time-code`.

**Comment by patrickhlauke** [comment]
i seem to remember that as far as Accessible Authentication is concerned, as long as you can copy/paste into the field for the one-time passcode (OTP), you satisfy the requirements of that SC

**Comment by patrickhlauke** [comment]
ah, i thought my edits to the understanding https://github.com/w3c/wcag/pull/1898 touched on OTPs as well, but they don't. might be something to add there...i'll have a think

**Comment by yatil** [comment]
> i seem to remember that as far as Accessible Authentication is concerned, as long as you can copy/paste into the field for the one-time passcode (OTP), you satisfy the requirements of that SC

Correct, hence I asked about extending the SC to require it 😉

**Comment by alastc** [comment]
Ah, that wasn't there when we did the WCAG 2.1 list, perhaps that something we can add. 

There's a task to review other new entries as well.

**Comment by patrickhlauke** [comment]
> perhaps that's something we can add

but again, not really a piece of info related to the individual...

**Comment by patrickhlauke** [comment]
> Correct, hence I asked about extending the SC to require it

note that 3.3.7 doesn't directly *require* the use of 1.3.5-specific input purpose identification. at its most basic, it requires that pasting and autofilling (if it works) isn't blocked, nothing more. i.e. you can *fail* 1.3.5, but still pass 3.3.7. at least that's my reading.

**Comment by alastc** [comment]
> a one-time code isn't really (personal) information related to the user

It doesn't have to be personal info, in this case it is something the user has, and does relate to the user. (No one else should have the same code at that time.)

**Comment by fstrr** [comment]
> There's a task to review other new entries as well.

I couldn't see an issue for this—if there is one, clearly I'm not searching for the right thing.

I think the only other new one I can see is `webauthn`, which is at the bottom of the table in the [autofill processing model section of the HTML spec](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#autofill-processing-model).

**Comment by alastc** [comment]
It isn't clear that this is "input", as it says "the user agent should show [public key credentials](https://w3c.github.io/webauthn/#public-key-credential) available via [conditional](https://w3c.github.io/webappsec-credential-management/#dom-credentialmediationrequirement-conditional) mediation when the user interacts with the form control".

I think we can leave that one.

**Comment by thw0rted** [comment]
I know I'm really late to this issue, but in response to

>  It doesn't have to be personal info, in this case it is something the user has, and does relate to the user

The [section in question](https://www.w3.org/WAI/WCAG21/Understanding/identify-input-purpose) begins with

> The intent of this Success Criterion is to ensure that the purpose of a form input collecting **information about the user** can be programmatically determined

(emphasis mine)

You could also say that what the user wants for lunch right now "relates to the user", but that doesn't make it information "about" the user.  Compare the concept of an ephemeral one-time-code, with a typical lifespan of 30 or 60 seconds, to every other value on the [accepted autocomplete value list](https://www.w3.org/TR/WCAG21/#input-purposes).  Everything else is either a long-lived property of the user's identity, or relates to credentials, contact details, residence, or financial info that is likely to remain unchanged for at least days to years.

I do think that WCAG should have an opinion about how to identify OTP fields, but IMHO they should fall into the bucket of "not about the user" and as such be tagged with something that identifies them as not suited for autofill.  (Now, if only we could [get the Chromium devs to listen](https://issues.chromium.org/issues/41239842) about respecting that tag...)

**Comment by mraccess77** [comment]
I don’t consider it to be about the user

**Comment by patrickhlauke** [comment]
@fstrr in light of the changes we recently proposed/made about `transaction-amount`, do you still feel that it's worth pursuing adding `one-time-code` to the list? i'm actually sway-able to squint a bit and say the OTP may be as related to the user as a password is, when looking at it in a certain light...

**Comment by fstrr** [comment]
sure, why not.

**Comment by mraccess77** [comment]
Seems like adding on-time-code feels like a normative change even if that section may not be considered normative.  Adding this could make things that were conformant before out of conformance.

**Comment by patrickhlauke** [comment]
i've softened to the (slightly stretched) idea that an one-time code is related to the user. Filed https://github.com/w3c/wcag/pull/5255

**Comment by thw0rted** [comment]
Maybe this is the wrong place to ask, but again... is this going to make it harder to convince browsers that one-time-code isn't suitable for autofill?  (You could add "current-heartrate" as "related to the user", but I'm not sure when the user-agent would benefit from including that in a normalized ontology of "purposes" either.)

**Comment by patrickhlauke** [comment]
@thw0rted not sure I understand your question. why would we want to "convince browsers"? `one-time-code` *is* an autofill value, and after some discussion here i've come over to the idea that it is - to a certain extent - "about the user", the same way that a password is. we're not here trying to influence what goes into the html spec, we're evaluating if what has been added/is there would be valuable to also include/mention in WCAG.

**Comment by thw0rted** [comment]
It's possible that I'm using the wrong language.  I alluded to a problem with browsers in my previous comment but realize I wasn't very clear or explicit.  The issue is with autofilling inappropriate values.  Chrome especially has [a very long history](https://issues.chromium.org/issues/41239842) of web devs complaining that the browser inserts saved values in forms where they don't belong (e.g.: when a nurse at a hospital is entering patient data, the browser autofills the address of the user/nurse), and Chrome has done little to address the problem as far as I can tell.  So when you say

> `one-time-code` is an autofill value

my worry is that browser developers will note that WCAG lists one-time-code as an autofill value, and use this as justification to populate saved (previously-entered) code values in the one-time-code field, which is obviously never the correct behavior.

Now, if the browser somehow has access to the "live" code (say, through OS level integration with a TOTP or SMS app) then I do see why standardizing the `autocomplete` attribute value makes sense, so the user-agent knows where to put the fresh code it has just obtained.  I may have been conflating that type of "autofill" behavior with the older stored-value-based fill behavior.  It's noteworthy that every other value in the list you updated in https://github.com/w3c/wcag/pull/5255 is "long lived" -- an attribute of the user that changes relatively infrequently.  It was alarming to see a short-lived value in the same place, so I just wanted to make sure the changed guidance didn't encourage the wrong thing.

**Comment by patrickhlauke** [comment]
honestly, i'd be surprised if browser developer cared at all about anything we do downstream in WCAG...
